```
!!! 'golint' problems: 
/home/travis/gopath/src/k8s.io/contrib/cluster-autoscaler/simulator/cluster.go:68:1: exported function CalculateReservation should have comment or be unexported
Found a flag declared with an _ but which is not explicitly listed as a valid flag name in hack/verify-flags/excluded-flags.txt
Are you certain this flag should not have been declared with an - instead?
scale_down_delay
scale_down_enabled
scale_down_trial_frequency
```
please fix

--- a/cluster-autoscaler/cluster_autoscaler.go
+++ b/cluster-autoscaler/cluster_autoscaler.go
@@ -30,8 +30,17 @@ import (
 )
 
 var (
-	migConfigFlag config.MigConfigFlag
-	kubernetes    = flag.String("kubernetes", "", "Kuberentes master location. Leave blank for default")
+	migConfigFlag    config.MigConfigFlag
+	kubernetes       = flag.String("kubernetes", "", "Kuberentes master location. Leave blank for default")
+	scaleDownEnabled = flag.Bool("experimental-scale-down-enabled", false, "Should CA scale down the cluster")
+	scaleDownDelay   = flag.Duration("scale-down-delay", 10*time.Minute,
+		"Duration from the last scale up to the time when CA starts to check scale down options")
+	scaleDownUnderutilizedTime = flag.Duration("scale-down-underutilized-time", 10*time.Minute,
+		"How long the node should be underutilized before it is eligible for scale down")
+	scaleDownUtilizationThreshold = flag.Float64("scale-down-utilization-threshold", 0.5,
+		"Node reservation level below which a node can be considered for scale down")
+	scaleDownTrialFrequency = flag.Duration("scale-down-trial-frequency", 10*time.Minute,
+		"How often scale down possiblity is check")
 )
 
 func main() {
@@ -43,15 +52,12 @@ func main() {
 	if err != nil {
 		glog.Fatalf("Failed to parse Kuberentes url: %v", err)
 	}
+
+	// Configuration
 	kubeConfig, err := config.GetKubeClientConfig(url)
 	if err != nil {
 		glog.Fatalf("Failed to build Kuberentes client configuration: %v", err)
 	}
-
-	kubeClient := kube_client.NewOrDie(kubeConfig)
-	unschedulablePodLister := NewUnschedulablePodLister(kubeClient)
-	nodeLister := NewNodeLister(kubeClient)
-
 	migConfigs := make([]*config.MigConfig, 0, len(migConfigFlag))
 	for i := range migConfigFlag {
 		migConfigs = append(migConfigs, &migConfigFlag[i])
@@ -62,7 +68,16 @@ func main() {
 		glog.Fatalf("Failed to create GCE Manager %v", err)
 	}
 
+	kubeClient := kube_client.NewOrDie(kubeConfig)
+
 	predicateChecker := simulator.NewPredicateChecker()
+	unschedulablePodLister := NewUnschedulablePodLister(kubeClient)
+	scheduledPodLister := NewScheduledPodLister(kubeClient)
+	nodeLister := NewNodeLister(kubeClient)
+
+	lastScaleUpTime := time.Now()
+	lastScaleDownFailedTrial := time.Now()
+	underutilizedNodes := make(map[string]time.Time)
 
 	for {
 		select {
@@ -97,12 +112,52 @@ func main() {
 
 				if len(unschedulablePodsToHelp) == 0 {
 					glog.V(1).Info("No unschedulable pods")
-					continue
+				} else {
+					scaledUp, err := ScaleUp(unschedulablePodsToHelp, nodes, migConfigs, gceManager, kubeClient, predicateChecker)
+					if err != nil {
+						glog.Errorf("Failed to scale up: %v", err)
+						continue
+					} else {
+						if scaledUp {
+							lastScaleUpTime = time.Now()
+							// No scale down in this iteration.
+							continue
+						}
+					}
 				}
 
-				_, err = ScaleUp(unschedulablePodsToHelp, nodes, migConfigs, gceManager, kubeClient, predicateChecker)
-				if err != nil {
-					glog.Errorf("Failed to scale up: %v", err)
+				if *scaleDownEnabled {
+
+					// In dry run only utilization is updated
+					calculateUtilizationOnly := lastScaleUpTime.Add(*scaleDownDelay).After(time.Now()) ||
+						lastScaleDownFailedTrial.Add(*scaleDownTrialFrequency).After(time.Now())
+
+					allScheduled, err := scheduledPodLister.List()
+					if err != nil {
+						glog.Errorf("Failed to list scheduled pods: %v", err)
+						continue
+					}
+					underutilizedNodes = CalculateUnderutilizedNodes(
+						nodes,
+						underutilizedNodes,
+						*scaleDownUtilizationThreshold,
+						allScheduled)
+
+					if !calculateUtilizationOnly {
+						result, err := ScaleDown(
+							nodes,
+							underutilizedNodes,
+							*scaleDownUnderutilizedTime,
+							allScheduled,
+							gceManager, kubeClient)
+						if err != nil {
+							glog.Errorf("Failed to scale down: %v", err)
+						} else {
+							if result != ScaleDownNodeDeleted {
+								lastScaleDownFailedTrial = time.Now()
+							}
+						}
+					}
 				}
 			}
 		}

--- a/cluster-autoscaler/scale_down.go
+++ b/cluster-autoscaler/scale_down.go
@@ -1,0 +1,133 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/contrib/cluster-autoscaler/config"
+	"k8s.io/contrib/cluster-autoscaler/simulator"
+	"k8s.io/contrib/cluster-autoscaler/utils/gce"
+	kube_api "k8s.io/kubernetes/pkg/api"
+	kube_client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+
+	"github.com/golang/glog"
+)
+
+// ScaleDownResult represents the state of scale down.
+type ScaleDownResult int
+
+const (
+	// ScaleDownError - scale down finished with error.
+	ScaleDownError ScaleDownResult = iota
+	// ScaleDownNoUnderutilized - no underutilized nodedes and no errors.
+	ScaleDownNoUnderutilized ScaleDownResult = iota
+	// ScaleDownNoNodeDeleted - underutilized nodes present but not available for deletion.
+	ScaleDownNoNodeDeleted ScaleDownResult = iota
+	// ScaleDownNodeDeleted - a node was deleted.
+	ScaleDownNodeDeleted ScaleDownResult = iota
+)
+
+// CalculateUnderutilizedNodes calculates which nodes are underutilized.
+func CalculateUnderutilizedNodes(nodes []*kube_api.Node,
+	underutilizedNodes map[string]time.Time,
+	utilizationThreshold float64,
+	pods []*kube_api.Pod) map[string]time.Time {
+
+	currentlyUnderutilizedNodes := make(map[string]struct{})
+	nodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
+
+	for _, node := range nodes {
+		nodeInfo, found := nodeNameToNodeInfo[node.Name]
+		if !found {
+			glog.Errorf("Node info for %s not found", node.Name)
+			continue
+		}
+		reservation, err := simulator.CalculateReservation(node, nodeInfo)
+
+		if err != nil {
+			glog.Warningf("Failed to calculate reservation for %s: %v", node.Name, err)
+		}
+		glog.V(4).Infof("Node %s - reservation %f", node.Name, reservation)
+
+		if reservation >= utilizationThreshold {
+			glog.V(4).Infof("Node %s is not suitable for removal - reservation to big (%f)", node.Name, reservation)
+			continue
+		}
+		currentlyUnderutilizedNodes[node.Name] = struct{}{}
+	}
+
+	now := time.Now()
+	result := make(map[string]time.Time)
+	for name := range currentlyUnderutilizedNodes {
+		if val, found := underutilizedNodes[name]; !found {
+			result[name] = now
+		} else {
+			result[name] = val
+		}
+	}
+	return result
+}
+
+// ScaleDown tries to scale down the cluster. It returns ScaleDownResult indicating if any node was
+// removed and error if such occured.
+func ScaleDown(
+	nodes []*kube_api.Node,
+	underutilizedNodes map[string]time.Time,
+	underutilizationTime time.Duration,
+	pods []*kube_api.Pod,
+	gceManager *gce.GceManager,
+	client *kube_client.Client) (ScaleDownResult, error) {
+
+	now := time.Now()
+	candidates := make([]*kube_api.Node, 0)
+	for _, node := range nodes {
+		if val, found := underutilizedNodes[node.Name]; found {
+			if val.Add(underutilizationTime).Before(now) {
+				candidates = append(candidates, node)
+			}
+		}
+	}
+	if len(candidates) == 0 {
+		glog.Infof("No candidates for scale down")
+		return ScaleDownNoUnderutilized, nil
+	}
+
+	nodeToRemove, err := simulator.FindNodeToRemove(candidates, nodes, pods, client)
+	if err != nil {
+		return ScaleDownError, fmt.Errorf("Find node to remove failed: %v", err)
+	}
+	if nodeToRemove == nil {
+		glog.V(1).Infof("No node to remove")
+		return ScaleDownNoNodeDeleted, nil
+	}
+	glog.Infof("Removing %s", nodeToRemove.Name)
+
+	instanceConfig, err := config.InstanceConfigFromProviderId(nodeToRemove.Spec.ProviderID)
+	if err != nil {
+		return ScaleDownError, fmt.Errorf("Failed to get instance config for %s: %v", nodeToRemove.Name, err)
+	}
+
+	err = gceManager.DeleteInstances([]*config.InstanceConfig{instanceConfig})
+	if err != nil {
+		return ScaleDownError, fmt.Errorf("Failed to delete %v: %v", instanceConfig, err)
+	}
+
+	return ScaleDownNodeDeleted, nil
+}

--- a/cluster-autoscaler/simulator/cluster.go
+++ b/cluster-autoscaler/simulator/cluster.go
@@ -29,36 +29,11 @@ import (
 	"github.com/golang/glog"
 )
 
-const (
-	// Nodes with utilization below this are considered unused and may be subject to scale down.
-	unusedThreshold = float64(0.5)
-)
-
 // FindNodeToRemove finds a node that can be removed.
-func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube_client.Client) (*kube_api.Node, error) {
-	nodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
+func FindNodeToRemove(candidates []*kube_api.Node, allNodes []*kube_api.Node, pods []*kube_api.Pod,
+	client *kube_client.Client) (*kube_api.Node, error) {
 
-	//TODO: Interate over underutulized nodes first.
-	for _, node := range nodes {
-		nodeInfo, found := nodeNameToNodeInfo[node.Name]
-		if !found {
-			glog.Errorf("Node info for %s not found", node.Name)
-			continue
-		}
-
-		reservation, err := calculateReservation(node, nodeInfo)
-
-		if err != nil {
-			glog.Warningf("Failed to calculate reservation for %s: %v", node.Name, err)
-		}
-		glog.V(4).Infof("Node %s - reservation %f", node.Name, reservation)
-
-		if reservation > unusedThreshold {
-			glog.Infof("Node %s is not suitable for removal - reservation to big (%f)", node.Name, reservation)
-			continue
-		}
-
-		// Let's try to remove this one.
+	for _, node := range candidates {
 		glog.V(2).Infof("Considering %s for removal", node.Name)
 
 		podsToRemoveList, _, _, err := cmd.GetPodsForDeletionOnNodeDrain(client, node.Name,
@@ -71,9 +46,9 @@ func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube
 
 		tempNodeNameToNodeInfo := schedulercache.CreateNodeNameToInfoMap(pods)
 		delete(tempNodeNameToNodeInfo, node.Name)
-		for _, node := range nodes {
-			if nodeInfo, found := tempNodeNameToNodeInfo[node.Name]; found {
-				nodeInfo.SetNode(node)
+		for _, tempnode := range allNodes {
+			if nodeInfo, found := tempNodeNameToNodeInfo[tempnode.Name]; found {
+				nodeInfo.SetNode(tempnode)
 			}
 		}
 		ptrPodsToRemove := make([]*kube_api.Pod, 0, len(podsToRemoveList))
@@ -81,7 +56,7 @@ func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube
 			ptrPodsToRemove = append(ptrPodsToRemove, &podsToRemoveList[i])
 		}
 
-		findProblems := findPlaceFor(ptrPodsToRemove, nodes, tempNodeNameToNodeInfo)
+		findProblems := findPlaceFor(ptrPodsToRemove, allNodes, tempNodeNameToNodeInfo)
 		if findProblems == nil {
 			return node, nil
 		}
@@ -90,7 +65,8 @@ func FindNodeToRemove(nodes []*kube_api.Node, pods []*kube_api.Pod, client *kube
 	return nil, nil
 }
 
-func calculateReservation(node *kube_api.Node, nodeInfo *schedulercache.NodeInfo) (float64, error) {
+// CalculateReservation calculates reservation of a node.
+func CalculateReservation(node *kube_api.Node, nodeInfo *schedulercache.NodeInfo) (float64, error) {
 	cpu, err := calculateReservationOfResource(node, nodeInfo, kube_api.ResourceCPU)
 	if err != nil {
 		return 0, err

--- a/cluster-autoscaler/simulator/cluster_test.go
+++ b/cluster-autoscaler/simulator/cluster_test.go
@@ -52,7 +52,7 @@ func TestReservation(t *testing.T) {
 			},
 		},
 	}
-	reservation, err := calculateReservation(node, nodeInfo)
+	reservation, err := CalculateReservation(node, nodeInfo)
 	assert.NoError(t, err)
 	assert.InEpsilon(t, 2.0/10, reservation, 0.01)
 
@@ -66,7 +66,7 @@ func TestReservation(t *testing.T) {
 			},
 		},
 	}
-	_, err = calculateReservation(node2, nodeInfo)
+	_, err = CalculateReservation(node2, nodeInfo)
 	assert.Error(t, err)
 }
 

--- a/hack/verify-flags/known-flags.txt
+++ b/hack/verify-flags/known-flags.txt
@@ -84,3 +84,8 @@ one-time
 change-permissions
 services-configmap
 use-real-cloud
+experimental-scale-down-enabled
+scale-down-delay
+scale-down-underutilized-time
+scale-down-utilization-threshold
+scale-down-trial-frequency

